File: Week_03/generate-parentheses.cpp
class Solution {
public:
    vector<string> res;

    vector<string> generateParenthesis(int n) {
        addParenthesis("", n, n);
        return res;
    }

    void addParenthesis(string s, int left, int right) {
        if (left == 0 && right == 0) {
            res.push_back(s);
            return;
        }
        if (left != 0) addParenthesis(s + "(", left - 1, right);
        if (right > left) addParenthesis(s + ")", left, right - 1);
    }
};

File: Week_03/generate-parentheses.py
class Solution:
    def __init__(self):
        self.res = []

    def generateParenthesis(self, n: int) -> List[str]:
        self.addParenthesis(n, n, '')
        return self.res

    def addParenthesis(self, left, right, s):
        if left == 0 and right == 0:
            self.res.append(s)

        if left > 0:    self.addParenthesis(left - 1, right, s + '(')
        if right > left:    self.addParenthesis(left, right - 1, s + ')')
